libs/game/sim/keymap.ts

```diff
@@ -0,0 +1,163 @@
+
+            // System keymap
+            this.setSystemKeys(
+                80, // P - Screenshot
+                82 // R - Gif
+            );
+
+            // Player 1 alternate mapping. This is cleared when the game sets any player keys explicitly
+            this.altmap[38] = Key.Up; // UpArrow
+            this.altmap[37] = Key.Left; // LeftArrow
+            this.altmap[40] = Key.Down; // DownArrow
+            this.altmap[39] = Key.Right; // RightArrow
+            this.altmap[81] = Key.A; // Q
+            this.altmap[90] = Key.A; // Z
+            this.altmap[88] = Key.B; // X
+            this.altmap[69] = Key.B; // E
+        }
+
+        public setPlayerKeys(
+            player: number, // player number is 1-based
+            up: number,
+            down: number,
+            left: number,
+            right: number,
+            A: number,
+            B: number
+        ) {
+            // We only support four players
+            if (player < 1 || player > 4) return;
+            const keyCodes = [up, down, left, right, A, B];
+            // Check for reserved key codes
+            // TODO: How to surface this runtime error to the user?
+            // TODO: Send message to UI: "Keyboard mapping contains a reserved key code"
+            const filtered = keyCodes.filter(keyCode => reservedKeyCodes.includes(keyCode));
+            if (filtered.length) return;
+            // Clear existing mapped keys for player
+            const mapName = `player-${player}`;
+            this.clearMap(mapName);
+            // Clear altmap When explicitly setting the player keys
+            this.altmap = {};
+            // Map the new keys
+            const offset = (player - 1) * 7; // +7 for player 2's keys
+            this.keymap[up] = Key.Up + offset;
+            this.keymap[down] = Key.Down + offset;
+            this.keymap[left] = Key.Left + offset;
+            this.keymap[right] = Key.Right + offset;
+            this.keymap[A] = Key.A + offset;
+            this.keymap[B] = Key.B + offset;
+            // Remember this mapping
+            this.saveMap(mapName, keyCodes);
+        }
+
+        public setSystemKeys(screenshot: number, gif: number) {
+            const mapName = "system";
+            // Clear existing mapped keys for system
+            this.clearMap(mapName);
+            this.keymap[screenshot] = Key.Screenshot;
+            this.keymap[gif] = Key.Gif;
+            // Remember this mapping
+            this.saveMap(mapName, [screenshot, gif]);
+        }
+
+        public getKey(keyCode: number): Key {
+            return this.keymap[keyCode] || this.altmap[keyCode] || Key.None;
+        }
+
+        private saveMap(name: string, keyCodes: number[]) {
+            this.mappings[name] = keyCodes;
+        }
+
+        private clearMap(name: string) {
+            const keyCodes = this.mappings[name];
+            keyCodes?.forEach(keyCode => delete this.keymap[keyCode]);
```

Comment:
I thought this runs alongside the static typescript compiler / gets mad at the `?.` feature of newer typescript? I guess if it works then I'm wrong though~

Reply:
This is in the sim.

Reply:
Does this run in the sim context, or STS? Now I'm not sure.

Reply:
The compiler expanded it to
```ts
keyCodes === null || keyCodes === void 0 ? void 0 : keyCodes.forEach(keyCode => delete this.keymap[keyCode]);
```

Reply:
So, it's called by the game (from inside the run/eval), but is not compiled by STS because it's a sim-only implementation of the api.

Reply:
Ah right, looks good -- I think last time I ran into that was on a branch that hadn't fully completed the bluebird change and it stuck in my head as it used to happen a lot.